### PR TITLE
Add nn.SwiGLU activation function

### LINKED ISSUE #128712: Add Swiglu activation function
### 🚀 The feature, motivation and pitch

Hey team, i love building things from scratch, and as i was implementing the LLaMa paper by meta obviously using pytorch i saw that pytorch did not have a nn.swiglu activation function. I did end up implementing it on my own, but i still feel like swiglu being used in mutliple other new language models, LLaMa being the most popler one, needs to be implemented in Pytorch and should be a part of it. 

### Alternatives

There are many other activation functions already in pytorch, but as results show, Swiglu has been really impactful in the success of LLaMa, so i think the community might benefit with it being in pytorch

### Additional context

Also, ive been using pytorch for a long time and ive always build things from scratch, and it has always been on my todo list to contribute to pytorch, i feel like this might be it.

### PR DESCRIPTION
**Pull request by saivedant169**
## Issue

Fixes #128712

## Summary

Adds `torch.nn.SwiGLU` and `torch.nn.functional.swiglu`, the SiLU-gated linear unit used in LLaMA, Mistral, Gemma, and most current transformer architectures.

```python
m = torch.nn.SwiGLU(dim=-1)
output = m(input)  # splits input in half, applies SiLU gating
```

The math: `SwiGLU(a, b) = SiLU(a) * b`, where input is split in half along `dim`.

Pure Python implementation using existing ops (`torch.chunk` + `torch.sigmoid`). The earlier attempt (#144465) went stale because it bundled C++ kernels, inductor lowering, and decompositions in one PR. This one keeps the surface small: get the API in first, leave a fused kernel for a follow-up.

Mirrors `nn.GLU`:
- Same class pattern (`__constants__`, `dim` parameter, `extra_repr`)
- Same functional API (scalar check, even-size check, `handle_torch_function`)
- Same docstring layout, with paper reference to [GLU Variants Improve Transformer](https://arxiv.org/abs/2002.05202)

Files touched:
- `torch/nn/modules/activation.py`: new `SwiGLU` class
- `torch/nn/functional.py`: new `swiglu()` function
- `torch/nn/modules/__init__.py`: re-export

## Checklist

- [ ] Passes lint (`spin fixlint`)
- [ ] Added/updated tests
- [ ] Updated documentation (if applicable)
- [ ] Included benchmark results (for PRs impacting perf)

## BC-breaking?

No. Adds a new public symbol, nothing existing changes behavior.

## Coordination

Commented on #128712 [here](https://github.com/pytorch/pytorch/issues/128712#issuecomment-4060966963).

### CODE REVIEW

**Comment by saivedant169** [comment]
@pytorch-bot label "release notes: feature"

**Comment by saivedant169** [comment]
Rebased on top of current main. No conflicts, same diff. Needs a maintainer to approve the workflow runs so CI can go green here.

@albanD @mikaylagawarecki @jbschlosser tagging for review when you have a moment, this mirrors the existing nn.GLU pattern so it should be small to look at.

@pytorchbot label "no-stale"

**Comment by saivedant169** [comment]
@albanD apologies for the direct ping. This one just needs a maintainer to approve the workflow runs so CI can go green, it has been stuck on that since I rebased on main. It is a small self-contained change, nn.SwiGLU with tests and docs. If you can kick off CI, or point me to whoever owns the nn activations these days, that would help a lot. Glad to adjust anything if something needs changing.

**Comment by chaObserv** [comment]
Although the order of gate(a) * b matches the paper, it causes inconsistency with the current implementation of nn.GLU (a * sigmoid(b)).

**Comment by saivedant169** [comment]
Right, the two don't agree on which half is the gate. I followed the GLU-variants paper (Shazeer 2020), where SwiGLU is Swish(xW) * xV, so the gate is the first half. nn.GLU gates the second half instead (a * sigmoid(b)). 

I'm fine either way: keep the paper order and call out the difference from nn.GLU in the docstring so it isn't surprising, or flip to a * SiLU(b) to match nn.GLU's convention. @albanD @jbschlosser any preference on which ships? Happy to change it.

**Comment by saivedant169** [comment]
@chaObserv went with keeping the paper order. SwiGLU in the paper is `Swish(xW) * xV` so the gate is the first half, and that's what LLaMA/Mistral/Gemma do. Flipping it to match `nn.GLU` would mean people get a SwiGLU that doesn't match the paper their model uses. Added a note in both docstrings so it's not a surprise if you're coming from `nn.GLU`.

@mikaylagawarecki this is the SwiGLU from #128712, Python impl using existing ops like you suggested there. CI has never actually run on it, the workflow runs need approving. Could you kick that off? Happy to change anything if needed.

**Comment by saivedant169** [comment]
@albanD this has been open since March and the workflow runs have never been approved, so there's nothing green here to review.

Rather than asking again, simpler question: does nn actually want SwiGLU? #128712 is labelled actionable and nn triage asked for a plain Python implementation using existing ops, which is what this is. If that's changed, say so and I'll close it. I'd rather know than keep bumping something nobody wants.

If it is wanted, it only needs someone to approve CI.

**Comment by erikhuck** [comment]
> @albanD this has been open since March and the workflow runs have never been approved, so there's nothing green here to review.
> 
> Rather than asking again, simpler question: does nn actually want SwiGLU? #128712 is labelled actionable and nn triage asked for a plain Python implementation using existing ops, which is what this is. If that's changed, say so and I'll close it. I'd rather know than keep bumping something nobody wants.
> 
> If it is wanted, it only needs someone to approve CI.

My team needs swiglu activation. So I can say that it is wanted by us

**Comment by erikhuck** [comment]
@saivedant169 It looks like the 2 checks have passed. Does that mean the workflow runs have been approved?

**Comment by saivedant169** [comment]
> @saivedant169 It looks like the 2 checks have passed. Does that mean the workflow runs have been approved?

Nah, those two aren't CI. EasyCLA is just the contributor agreement bot, and the Meta one checks whether the PR touches internal-only files. Neither runs any code.
